File: python/0006_ZigZag Conversion.py
# The string "PAYPALISHIRING" is written in a zigzag pattern on a given number of rows like this: (you may want to display this pattern in a fixed font for better legibility)

# P   A   H   N
# A P L S I I G
# Y   I   R
# And then read line by line: "PAHNAPLSIIGYIR"

# Write the code that will take a string and make this conversion given a number of rows:

# string convert(string s, int numRows);

# Example 1:
# Input: s = "PAYPALISHIRING", numRows = 3
# Output: "PAHNAPLSIIGYIR"

# Example 2:
# Input: s = "PAYPALISHIRING", numRows = 4
# Output: "PINALSIGYAHRPI"
# Explanation:

# P     I    N
# A   L S  I G
# Y A   H R
# P     I

class Solution:
    def convert(self, s: str, numRows: int) -> str:
        length = len(s)
        if length <= numRows or numRows == 1:
            return s

        result = ''

        for row in range(numRows):
            step = [2 * (numRows - row - 1), 2 * row]
            current_step = 0
            start_pos = row
            result += s[start_pos]
            while start_pos + step[current_step % 2] < length:
                result += ('' 
                    if step[current_step % 2] == 0 else 
                    s[start_pos + step[current_step % 2]]
                )
                start_pos += step[current_step % 2]
                current_step += 1
        
        return result


if __name__ == "__main__":
    s = Solution()
    print(s.convert('PAYPALISHIRING', 10))

File: 0002-0050/0006_ZigZag Conversion.js
/** ***************************
  [Description]
  The string "PAYPALISHIRING" is written in a zigzag pattern on a given number of rows like this:
  P   A   H   N
  A P L S I I G
  Y   I   R

  [Input]
  s = "PAYPALISHIRING", numRows = 4

  [Output]
  "PINALSIGYAHRPI"

  [Explanation]
  P     I    N
  A   L S  I G
  Y A   H R
  P     I

  [My Method]
  P A Y P A L I S H I R I N G
  0           0           0
    1       1   1       1   1
      2   2       2   2
        3           3

  0 -> P     I     N
  1 -> A   L S   I G
  2 -> Y A   H R
  3- > P     I
*************************** */

/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
function convert(s, numRows) {
  if (numRows === 1 || s.length < 2 || numRows >= s.length) {
    return s;
  }
  let result = '';
  const lengthS = s.length;
  const step = 2 * (numRows - 1);
  const times = Math.ceil(lengthS / numRows);
  for (let i = 0; i < numRows; i++) {
    for (let j = 0; j < times; j++) {
      const position1 = i + j * step;
      const position2 = (j + 1) * step - i;
      if (i !== numRows - 1 && position1 < lengthS) {
        result += s[position1];
      }
      if (i !== 0 && position2 < lengthS) {
        result += s[position2];
      }
    }
  }
  return result;
}

function main() {
  console.log(convert('ABC', 2));
}

main();
